From 0d5445a0186ee8db7d37ebae84c42f4b9966de96 Mon Sep 17 00:00:00 2001
From: Claudio Jeker <claudio@openbsd.org>
Date: Thu, 14 Feb 2019 11:51:05 +0100
Subject: [PATCH 1/6] depend on carp requieres HAVE_CARP

---
 src/usr.sbin/bgpd/parse.y | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git src/usr.sbin/bgpd/parse.y src/usr.sbin/bgpd/parse.y
index 29af28a313..17b74ac759 100644
--- src/usr.sbin/bgpd/parse.y
+++ src/usr.sbin/bgpd/parse.y
@@ -2274,6 +2274,7 @@ peeropts	: REMOTEAS as4number	{
 			free($3);
 		}
 		| DEMOTE STRING		{
+#ifdef HAVE_CARP
 			if (strlcpy(curpeer->conf.demote_group, $2,
 			    sizeof(curpeer->conf.demote_group)) >=
 			    sizeof(curpeer->conf.demote_group)) {
@@ -2290,6 +2291,11 @@ peeropts	: REMOTEAS as4number	{
 				    curpeer->conf.demote_group);
 				YYERROR;
 			}
+#else
+			yyerror("carp demote not supported");
+			free($2);
+			YYERROR;
+#endif
 		}
 		| TRANSPARENT yesno	{
 			if ($2 == 1)
-- 
2.46.0

