From fec370a8b49c290563df245324192d2f8524ecdb Mon Sep 17 00:00:00 2001
From: Claudio Jeker <claudio@openbsd.org>
Date: Thu, 14 Feb 2019 11:51:05 +0100
Subject: [PATCH 1/5] depend on carp requieres HAVE_CARP

---
 src/usr.sbin/bgpd/parse.y | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git src/usr.sbin/bgpd/parse.y src/usr.sbin/bgpd/parse.y
index 9f5df25ba1..f3ee180ade 100644
--- src/usr.sbin/bgpd/parse.y
+++ src/usr.sbin/bgpd/parse.y
@@ -1617,6 +1617,7 @@ peeropts	: REMOTEAS as4number	{
 			free($3);
 		}
 		| DEMOTE STRING		{
+#ifdef HAVE_CARP
 			if (strlcpy(curpeer->conf.demote_group, $2,
 			    sizeof(curpeer->conf.demote_group)) >=
 			    sizeof(curpeer->conf.demote_group)) {
@@ -1633,6 +1634,11 @@ peeropts	: REMOTEAS as4number	{
 				    curpeer->conf.demote_group);
 				YYERROR;
 			}
+#else
+			yyerror("carp demote not supported");
+			free($2);
+			YYERROR;
+#endif
 		}
 		| TRANSPARENT yesno	{
 			if ($2 == 1)
-- 
2.26.1

