From 6a435ac7c314229981d56d2eb05f8f661152969c Mon Sep 17 00:00:00 2001
From: Claudio Jeker <claudio@openbsd.org>
Date: Fri, 12 Apr 2019 17:17:08 +0200
Subject: [PATCH 4/5] Allow overriding default user and file locations

---
 src/usr.sbin/bgpd/bgpd.h | 16 +++++++++++++---
 1 file changed, 13 insertions(+), 3 deletions(-)

diff --git src/usr.sbin/bgpd/bgpd.h src/usr.sbin/bgpd/bgpd.h
index f56a320c6e..68e3a1c32d 100644
--- src/usr.sbin/bgpd/bgpd.h
+++ src/usr.sbin/bgpd/bgpd.h
@@ -35,8 +35,6 @@
 
 #define	BGP_VERSION			4
 #define	BGP_PORT			179
-#define	CONFFILE			"/etc/bgpd.conf"
-#define	BGPD_USER			"_bgpd"
 #define	PEER_DESCR_LEN			32
 #define	REASON_LEN			256	/* includes NUL terminator */
 #define	PFTABLE_LEN			32
@@ -70,7 +68,19 @@
 
 #define	BGPD_LOG_UPDATES		0x0001
 
-#define	SOCKET_NAME			"/var/run/bgpd.sock"
+#ifndef BGPD_USER
+#define	BGPD_USER			"_bgpd"
+#endif
+
+#ifndef SYSCONFDIR
+#define SYSCONFDIR "/etc"
+#endif
+#define	CONFFILE			SYSCONFDIR "/bgpd.conf"
+
+#ifndef RUNSTATEDIR
+#define RUNSTATEDIR "/var/run"
+#endif
+#define	SOCKET_NAME			RUNSTATEDIR "/bgpd.sock"
 
 #define	F_BGPD_INSERTED		0x0001
 #define	F_KERNEL		0x0002
-- 
2.31.1

